Adjacent

kmoham6 opened the pull request:
Adapting adjacent_difference to C++20

hkaiser reviewed (commented):
Could you please update the docs as well: 
- https://github.com/STEllAR-GROUP/hpx/blob/master/docs/sphinx/api/public_api.rst and 
- https://github.com/STEllAR-GROUP/hpx/blob/master/docs/sphinx/manual/writing_single_node_hpx_applications.rst?

hkaiser commented:
Also, please fix the compilation problems exposed by the CI.

Jedi18 commented on libs/core/include_local/include/hpx/local/numeric.hpp:
```diff
@@ -10,5 +10,8 @@
 
 namespace hpx {
-    using hpx::parallel::adjacent_difference;
+    using hpx::parallel::exclusive_scan;
```
Could you remove these since they have been removed on master?

hkaiser commented:
@kmoham6 could you please take care of the remaining CI problems?

kmoham6 commented:
@hkaiser I am working with Srinivas. I am taking care of the changed files in my branch. conflict will be fixed soon.

msimberg commented on libs/core/algorithms/include/hpx/parallel/algorithms/detail/adjacent_difference.hpp:
```diff
@@ -27,13 +28,24 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
+            while (++first != last)
+            {
+                value_t val = *first;
+                *++dest = op(val, std::move(acc));    // std::move since C++20
```
```suggestion
                *++dest = HPX_INVOKE(op, val, std::move(acc));
```
What is the `C++20` comment saying?

kmoham6 replied:
I removed this C++20 comment

msimberg commented on libs/core/algorithms/tests/unit/algorithms/adjacentdifference.cpp:
```diff
@@ -24,48 +26,6 @@ void adjacent_difference_test()
-{
-    using namespace hpx::execution;
-
-    // If the execution policy object is of type vector_execution_policy,
-    // std::terminate shall be called. therefore we do not test exceptions
-    // with a vector execution policy
-    test_adjacent_difference_exception(seq, IteratorTag());
-    test_adjacent_difference_exception(par, IteratorTag());
-
-    test_adjacent_difference_exception_async(seq(task), IteratorTag());
-    test_adjacent_difference_exception_async(par(task), IteratorTag());
-}
-
-void adjacent_difference_exception_test()
-{
-    test_adjacent_difference_exception<std::random_access_iterator_tag>();
-    test_adjacent_difference_exception<std::forward_iterator_tag>();
-}
-
-template <typename IteratorTag>
-void test_adjacent_difference_bad_alloc()
-{
-    using namespace hpx::execution;
-
-    // If the execution policy object is of type vector_execution_policy,
-    // std::terminate shall be called. therefore we do not test exceptions
-    // with a vector execution policy
-    test_adjacent_difference_bad_alloc(seq, IteratorTag());
-    test_adjacent_difference_bad_alloc(par, IteratorTag());
-
-    test_adjacent_difference_bad_alloc_async(seq(task), IteratorTag());
-    test_adjacent_difference_bad_alloc_async(par(task), IteratorTag());
-}
-
-void adjacent_difference_bad_alloc_test()
-{
-    test_adjacent_difference_bad_alloc<std::random_access_iterator_tag>();
-    test_adjacent_difference_bad_alloc<std::forward_iterator_tag>();
-}
-
```
Have these moved somewhere else or are they no longer relevant?

kmoham6 replied:
Yes, they have moved to the files below: 
libs/core/algorithms/tests/unit/algorithms/adjacentdifference_exception.cpp
libs/core/algorithms/tests/unit/algorithms/adjacentdifference_bad_alloc.cpp

msimberg commented on libs/core/include_local/include/hpx/local/numeric.hpp:
```diff
@@ -10,5 +10,5 @@
 namespace hpx {
-    using hpx::parallel::adjacent_difference;
+
 }    // namespace hpx
```
You can remove the whole `namespace hpx {}` here.

kmoham6 replied:
I removed whole "namespace hpx {}"

hkaiser commented:
@kmoham6 the cmake_format issue should be gone if you rebase on master. The other test issues are unrelated and known, you will not have to worry about those.

msimberg commented:
@kmoham6 could you mark comments that you've already taken care of as resolved so that we can see what is actually still unresolved?

kmoham6 commented on libs/core/algorithms/include/hpx/parallel/algorithms/adjacent_difference.hpp:
```diff
@@ -16,7 +154,9 @@
 #include <hpx/executors/execution_policy.hpp>
 #include <hpx/parallel/algorithms/detail/adjacent_difference.hpp>
```
I changed the name to adjacent_difference.hpp

hkaiser commented:
@kmoham6 not sure if I missed it before, but now we see compilation problems with the datapar tests: https://cdash.cscs.ch/viewBuildError.php?buildid=193713

kmoham6 commented:
@hkaiser seems the error is about std::minus<> that I changed it before my last rebase. it looks like they were lost after rebase. I fixed them and pushing again.

kmoham6 commented:
> @kmoham6 not sure if I missed it before, but now we see compilation problems with the datapar tests: https://cdash.cscs.ch/viewBuildError.php?buildid=193713

hkaiser reviewed (approved):
This is now ready to go in. LGTM, thanks!
